crates/bevy_ecs/src/bundle.rs

```diff
@@ -571,192 +501,379 @@ impl BundleInfo {
     }
 }
 
-pub(crate) struct BundleInserter<'a, 'b> {
-    pub(crate) archetype: &'a mut Archetype,
-    pub(crate) entities: &'a mut Entities,
-    bundle_info: &'b BundleInfo,
-    table: &'a mut Table,
-    sparse_sets: &'a mut SparseSets,
-    result: InsertBundleResult<'a>,
-    archetypes_ptr: *mut Archetype,
+// SAFETY: We have exclusive world access so our pointers can't be invalidated externally
+pub(crate) struct BundleInserter<'w> {
+    world: UnsafeWorldCell<'w>,
+    bundle_info: *const BundleInfo,
```

Comment:
Having these be `NonNull` would be ideal to verify it's OK to at least dereference them.

Reply:
I can change them all to `NonNull` but then we lose the const/mut-ness, either way we have a hidden invariant that we just promise we won't break. We could also just do NonNull for the mutable pointers. Not sure what's best here so will also go with anything.

Reply:
Probably best to leave it then. Might be a good idea to add mut/const variants of `NonNull` to bevy_ptr as follow-up.